### PR TITLE
Feature/breakpoints

### PR DESCRIPTION
**Pull request by MatthewCMoya**
* Responsive feature
* unit tests
* global error class
* timing issue with onCreate - beforeCreate

### CODE REVIEW

File: src/Ad.ts [hunk 1]
```diff
@@ -204,7 +192,10 @@ class Ad implements IAd {
     this.container = insertElement('div', { style: 'position: relative; display: inline-block;' }, el);
     this.el = insertElement('div', { id: nextId() }, this.container);
 
-    this.networkInstance = this.network.createAd(this);
+    // TODO: add this in when writing tests
```
Comment (MatthewCMoya) [inline src/Ad.ts]: I will uncomment this before merge

File: src/Ad.ts [hunk 2]
```diff
@@ -215,10 +205,21 @@ class Ad implements IAd {
...
 
-    this.callPlugins('onCreate');
+    const executionOfPlugins = this.callPlugins('beforeCreate');
 
     this.onReady(
-      () => this.callPlugins('afterCreate'),
+      async () => {
+        await executionOfPlugins;
+
+        await Promise.all(
```
Comment (MatthewCMoya) [inline src/Ad.ts]: This should not be a promise.all. It will be far too brittle for the use case.
I will create a work item to address it in a following pr

File: src/Ad.ts [hunk 3]
```diff
@@ -58,7 +48,7 @@ function attachAsLifecycleMethod(
     if (this.state.frozen) {
       const boundReplayFn = this[propertyName].bind(this, ...args);
 
-      this.actionsReceievedWhileFrozen.push(boundReplayFn);
+      this.actionsreceivedWhileFrozen.push(boundReplayFn);
```
Comment (stephenbaldwin) [inline src/Ad.ts]: `actionsReceivedWhileFrozen`

File: src/Ad.ts [hunk 4]
```diff
@@ -165,9 +153,9 @@ class Ad implements IAd {
 
   public configuration: IAdConfiguration;
 
-  private networkInstance: INetworkInstance;
+  private networkInstance!: INetworkInstance;
 
-  private actionsReceievedWhileFrozen: any = [];
+  private actionsreceivedWhileFrozen: any = [];
```
Comment (stephenbaldwin) [inline src/Ad.ts]: actionsReceivedWhileFrozen

File: src/Ad.ts [hunk 5]
```diff
@@ -319,9 +320,9 @@ class Ad implements IAd {
     // the queue for this event.
     this.state.frozen = false;
 
-    const actions = this.actionsReceievedWhileFrozen;
+    const actions = this.actionsreceivedWhileFrozen;
```
Comment (stephenbaldwin) [inline src/Ad.ts]: actionsReceivedWhileFrozen

File: src/Ad.ts [hunk 6]
```diff
@@ -319,9 +320,9 @@ class Ad implements IAd {
...
 
-    this.actionsReceievedWhileFrozen = [];
+    this.actionsreceivedWhileFrozen = [];
```
Comment (stephenbaldwin) [inline src/Ad.ts]: actionsReceivedWhileFrozen

File: src/plugins/Breakpoints.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import isBetween from '../utils/isBetween';
+import throttle from '../utils/throttle';
+import GenericPlugin from './GenericPlugin';
+
+class BreakpointsPlugin extends GenericPlugin {
+  public currentConfines: number[] = [];
+  private EVENT_KEY: string = 'resize';
+  private THROTTLE_DURATION: number = 200;
+
+  public beforeCreate(ad: IAd) {
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: instead of using the param ad use `this.ad`

File: src/plugins/Breakpoints.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        return;
+      }
+
+      console.log('Entered new breakpoint. refreshing ad...');
+      this.determineCurrentBreakpoint(breakpoints);
+      ad.refresh();
+    }));
+  }
+
+  public determineCurrentBreakpoint(breakpoints: IAdBreakpoints) {
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: no need to pass in breakpoints here. you can grab them via `this.ad. configuration.breakpoints`

File: src/plugins/Breakpoints.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import throttle from '../utils/throttle';
+import GenericPlugin from './GenericPlugin';
+
+class BreakpointsPlugin extends GenericPlugin {
+  public currentConfines: number[] = [];
+  private EVENT_KEY: string = 'resize';
+  private THROTTLE_DURATION: number = 200;
+
+  public beforeCreate() {
+    this.determineCurrentBreakpoint();
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: We're missing the check for `refreshOnBreakpoint`.

Documentation here: https://adjs.dev/#/responsive-plugin?id=options

File: src/plugins/Breakpoints.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      }
+    });
+    if (!breakpointSpecifiedForViewWidth) {
+      this.currentConfines = [];
+      console.warn(`No breakpoints specified for display: ${viewWidth}px`);
+    }
+  }
+}
+
+export default BreakpointsPlugin;
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: You'll want to change this to `export = BreakpointsPlugin;` due to the last merge

File: src/plugins/Breakpoints.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    let breakpointSpecifiedForViewWidth: boolean = false;
+
+    Object.keys(breakpoints).forEach((key) => {
+      const { from, to } = breakpoints[key];
+      if (isBetween(viewWidth, from, to)) {
+        breakpointSpecifiedForViewWidth = true;
+        this.currentConfines = [from, to];
+      }
+    });
+    if (!breakpointSpecifiedForViewWidth) {
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: let's add a space above this

File: src/plugins/Breakpoints.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    if (!breakpoints) {
+      throw new AdJsError('MISCONFIGURATION', 'Breakpoints must be provided for BreakpointsPlugin use.');
+    }
+
+    const viewWidth: number = window.innerWidth;
+    let breakpointSpecifiedForViewWidth: boolean = false;
+
+    Object.keys(breakpoints).forEach((key) => {
+      const { from, to } = breakpoints[key];
+      if (isBetween(viewWidth, from, to)) {
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: and a space above this

File: src/plugins/Breakpoints.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        breakpointSpecifiedForViewWidth = true;
+        this.currentConfines = [from, to];
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: Should we store this as an object keyed with "from" and "to" so that we're better able to read the lines above?

i.e.
```
if (isBetween(window.innerWidth, this.currentConfines.from, this.currentConfines.to)) {
```

instead of
```
if (isBetween(window.innerWidth, this.currentConfines[0], this.currentConfines[1])) {
```

File: src/plugins/Breakpoints.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+  public beforeCreate() {
+    this.determineCurrentBreakpoint();
+
+    window.addEventListener(this.EVENT_KEY, throttle(this.THROTTLE_DURATION, () => {
+      if (isBetween(window.innerWidth, this.currentConfines[0], this.currentConfines[1])) {
+        return;
+      }
+
+      console.log('Entered new breakpoint. refreshing ad...');
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: Let's kill the console logs. We'll build a message system that links into the "logging plugin"

File: src/plugins/Breakpoints.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    Object.keys(breakpoints).forEach((key) => {
+      const { from, to } = breakpoints[key];
+      if (isBetween(viewWidth, from, to)) {
+        breakpointSpecifiedForViewWidth = true;
+        this.currentConfines = [from, to];
+      }
+    });
+    if (!breakpointSpecifiedForViewWidth) {
+      this.currentConfines = [];
+      console.warn(`No breakpoints specified for display: ${viewWidth}px`);
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: Let's kill the logging here as well. Next iteration will link into the logging plugin

File: src/plugins/Breakpoints.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+      this.determineCurrentBreakpoint();
+      this.ad.refresh();
+    }));
+  }
+
+  public determineCurrentBreakpoint() {
+    const { breakpoints } = this.ad.configuration;
+
+    if (!breakpoints) {
+      throw new AdJsError('MISCONFIGURATION', 'Breakpoints must be provided for BreakpointsPlugin use.');
```
Comment (stephenbaldwin) [inline src/plugins/Breakpoints.ts]: Just return if no breakpoints provided. This library should error if you've provided sizes in an object form __without__ breakpoints, however that should be done on instantiation in the ad class itself

File: src/Ad.ts [hunk 17]
```diff
@@ -204,7 +192,9 @@ class Ad implements IAd {
     this.container = insertElement('div', { style: 'position: relative; display: inline-block;' }, el);
     this.el = insertElement('div', { id: nextId() }, this.container);
 
-    this.networkInstance = this.network.createAd(this);
+    if (!Array.isArray(this.configuration.sizes) && !this.configuration.breakpoints) {
+      throw new AdJsError('MISCONFIGURATION', 'Sizes must be of type `Array` without BreakpointPlugin');
```
Comment (stephenbaldwin) [inline src/Ad.ts]: I would reword this due to our last modification.

```
Sizes must be of type `Array` unless breakpoints have been specified 
```

File: src/Ad.ts [hunk 18]
```diff
@@ -204,7 +192,9 @@ class Ad implements IAd {
     this.container = insertElement('div', { style: 'position: relative; display: inline-block;' }, el);
     this.el = insertElement('div', { id: nextId() }, this.container);
 
-    this.networkInstance = this.network.createAd(this);
+    if (!Array.isArray(this.configuration.sizes) && !this.configuration.breakpoints) {
```
Comment (stephenbaldwin) [inline src/Ad.ts]: Maybe break this conditional into a function for readability?

```
function sizesAreValid(this.configuration) {
}

if (!sizesAreValid()) {
}
```

File: src/Ad.ts (hunk 2 above)
Comment (stephenbaldwin) [reply src/Ad.ts]: 👍 Good catch

File: src/types.ts [hunk 19]
```diff
@@ -59,6 +59,7 @@ export type IPluginConstructor = new(ad: IAd) => IPlugin;
 export interface IPlugin {
   name: string;
 
+  beforeCreate?: IPluginHook;
   onCreate?: IPluginHook;
   afterCreate?: IPluginHook;
```
Comment (MatthewCMoya) [inline src/types.ts]: Do we need to add beforeDestroy() and beforeClear() here?
